### PR TITLE
[Merged by Bors] - Cascaded shadow maps.

### LINKED ISSUE #3629: Cascaded Shadow Maps
# Description

- Directional light shadow maps have quality pretty heavily bound by the ratio of shadow map texels to screen fragments.
  - If the relative resolution of the directional light shadow map is low, then its texels cover multiple screen fragments when they are viewed using the light's projection. This means that multiple screen fragments read from the same shadow map texel which results in aliasing.
  - Using a large shadow map texture is both expensive to process and uses a lot of VRAM
  - In order to have a ratio of shadow map texels to screen fragments approaching 1:1 to get good visual quality when fragments are close to the camera, the ratio then becomes very large for fragments further away from the camera which means that a lot of VRAM has to be wasted to achieve good visual quality close to the camera.

## Solution

- Cascaded shadow maps - split the depth range over which shadows are visible into overlapping slices (cascades) with orthographic projections fit to cover the depth slice, with a far plane at least behind all visible geometry and a near plane ideally slightly closer to the light than the geometry that is closest to the light. When sampling fragments that fall in regions covered by two cascades, use a weighted blend of shadow samples from both cascades to transition smoothly from one to the next in order to minimise visible artifacts as the ratio of shadow map texels to screen fragments changes.
- https://github.com/superdump/bevy/tree/cascade-shadow-maps
  - An implementation of the above that fits the cascade projections only to the _visible_ geometry, and not also the near plane to the geometry that is not visible, but that is closer to the light source than the visible geometry.

## Next Steps

- Rebase the branch
- Implement proper shadow projection fitting as documented in _Foundations of Game Engine Development 2: Rendering_ by Eric Lengyel, which is a reference book I have been using for much of the frustum culling and some of the shadow mapping that I've implemented so far.
- Make a PR

### PR DESCRIPTION
**Pull request by danchia**
Co-authored-by: Robert Swain <robert.swain@gmail.com>

# Objective

Implements cascaded shadow maps for directional lights, which produces better quality shadows without needing excessively large shadow maps.

Fixes #3629

Before
image

After
image

## Solution

Rather than rendering a single shadow map for directional light, the view frustum is divided into a series of cascades, each of which gets its own shadow map. The correct cascade is then sampled for shadow determination.

---

## Changelog

Directional lights now use cascaded shadow maps for improved shadow quality.

## Migration Guide

You no longer have to manually specify a `shadow_projection` for a directional light, and these settings should be removed. If customization of how cascaded shadow maps work is desired, modify the `CascadeShadowConfig` component instead.

### CODE REVIEW

**Comment by mockersf** [comment]
Nice!

Your branch is 78 commits late from main and has conflicts, could you resolve those?

**Comment by danchia** [comment]
> Nice!
> 
> Your branch is 78 commits late from main and has conflicts, could you resolve those?

Rebased + fixed a bug to do with shader indexing.

File: examples/tools/scene_viewer/scene_viewer_plugin.rs
Comment (superdump) [inline examples/tools/scene_viewer/scene_viewer_plugin.rs:222]: The instructions also need to be removed from the `impl fmt::Display for SceneHandle`.

**Review by superdump — CHANGES_REQUESTED** [review]
Initial pass.

**Comment by superdump** [comment]
Performance seems to be very poor on this PR with shadows enabled even on a fairly simple scene like the flight helmet using the `scene_viewer` example without any arguments.

**Comment by danchia** [comment]
> 

Interesting, I will take a closer look as to why

File: crates/bevy_pbr/src/render/light.rs
Comment (danchia) [reply crates/bevy_pbr/src/render/light.rs:1206]: There is a sort above that sorts shadow-enabled lights first

**Comment by superdump** [comment]
As a note, this is on an M1 Max. I'm continuing investigations on Discord as it will be noisy to spam random findings here. :)

**Comment by superdump** [comment]
With changes like those in #7069 also for the scope of this PR we were able to keep performance in check. :)

File: crates/bevy_pbr/src/light.rs
Comment (danchia) [reply crates/bevy_pbr/src/light.rs:317]: Agreed that it will be as large close as far, but it may still make sense to devote more pixels to closer distances than further ones w.r.t to quality bang for buck :)

File: examples/tools/scene_viewer/scene_viewer_plugin.rs
Comment (danchia) [reply examples/tools/scene_viewer/scene_viewer_plugin.rs:222]: Done.

File: crates/bevy_pbr/src/render/mesh_view_types.wgsl
Comment (danchia) [reply crates/bevy_pbr/src/render/mesh_view_types.wgsl]: Done

**Comment by danchia** [comment]
I made changes to the shader to follow what https://github.com/bevyengine/bevy/pull/7069 did specifically for directional shadowing code that I modified, left the rest of the optimizations to that PR.

**Comment by superdump** [comment]
I'll give this another review pass when I can.

**Comment by mockersf** [comment]
I tried comparing example `lighting` for main and this pr

https://user-images.githubusercontent.com/8672791/210342283-8e1aaa7b-90a3-45d7-9a7a-620d470915ac.mp4

Shadow borders are sometime less clean, and there's something strange happening in the back at 5 seconds

**Comment by superdump** [comment]
Good catch @mockersf ! The aliased edges is a cascade bound tweaking issue. By default we're configuring 4 cascades with the first cascade far bound at 5 units, and the last at 1000 units. This scene is small so 1000 units is way too much. But also we probably just want to adjust a bit to something that looks good out of the box with the current 2x2 Percentage Closer Filtering (using multiple samples of the shadow map to filter the edges).

However, the bug in the back behind the wall is something we will need to fix. It looks like the red and blue light are making it through the wall...
Screenshot 2023-01-03 at 12 05 01
The red and blue lights are point lights.

**Comment by mockersf** [comment]
WASM also has something funny:

https://user-images.githubusercontent.com/8672791/210359269-48057ce0-d50e-44b3-9e4b-0dfd257069b0.mp4

**Comment by danchia** [comment]
> WASM also has something funny:
> 
>  cascaded-wasm.mp4

Ah, for WASM (which I assume uses the `webgl` feature) we should probably change the default cascade config to only 1 cascade, and some more reasonable far bound that 1000. webgl doesn't have texture arrays, so we can only use one cascade.

But I think WASM users are going to have to manually configure the far bound for best results still. since achieving a tight bound is important for quality vs shadow map resolution.

**Comment by danchia** [comment]
I will look into what's going on with the point lights bleeding through the wall.

**Comment by danchia** [comment]
* Fixed the weird shadow light bleeding problems. Turns out my depth clamp (as written) only works on ortho matrices, so I changed the code to only turn on the clamping for directional lights.
* Updating lighting example to show how / when to customize the cascades config
* Updated the default cascades config for webgl to only use 1 cascade, covering a smaller shadow distance.


https://user-images.githubusercontent.com/1222141/211190615-4a1faf5d-f885-4840-a799-4dac6a67a3d6.mp4

File: crates/bevy_pbr/src/render/shadows.wgsl
Comment (superdump) [inline crates/bevy_pbr/src/render/shadows.wgsl:169]:
```suggestion
    // Blend with the next cascade, if there is one.
```

File: examples/3d/load_gltf.rs
Comment (superdump) [inline examples/3d/load_gltf.rs:25]: Probably also tweak the cascades here so this one looks good. It's kind of the fanciest-looking in-repo PBR example.

**Review by superdump — COMMENTED** [review]
This looks pretty good to me. I'm going to do a bit more testing locally but my finger is hovering over approval. We can rebase after #6284 is merged.

**Comment by rparrett** [comment]
Should there be a warning or something when `CascadeShadowConfig` has `num_cascades` > `MAX_CASCADES_PER_LIGHT`?

In a test scene where the camera was pretty far from the action, bumping value up that resulted in shadows mysteriously disappearing (as only the first 4 cascades were used?).

**Comment by danchia** [comment]
> Should there be a warning or something when `CascadeShadowConfig` has `num_cascades` > `MAX_CASCADES_PER_LIGHT`?
> 
> In a test scene where the camera was pretty far from the action, bumping value up that resulted in shadows mysteriously disappearing (as only the first 4 cascades were used?).

That's a good idea, will look into it.

File: crates/bevy_pbr/src/light.rs
Comment (robtfm) [inline crates/bevy_pbr/src/light.rs:281]: please can we assert `overlap_proportion < 1` as well? >1 causes it to hit the z_near assert below but is a little more cryptic.

File: examples/tools/scene_viewer/scene_viewer_plugin.rs
Comment (robtfm) [inline examples/tools/scene_viewer/scene_viewer_plugin.rs:47]: could we keep some inputs for updating the cascade far and near settings?

File: crates/bevy_render/src/view/mod.rs
Comment (robtfm) [inline crates/bevy_render/src/view/mod.rs:90]: would it be possible to remove the old projection field and just do the projection * inverse calculation using `view_projection` wherever `projection` is currently used? then you have the fine control you need without duplicating the data?

if not then can we add some comments here. it seems like a source of potential future confusion to have a `projection` and a `view_projection` field with no descriptions.

**Review by robtfm — CHANGES_REQUESTED** [review]

File: examples/tools/scene_viewer/scene_viewer_plugin.rs
Comment (robtfm) [reply examples/tools/scene_viewer/scene_viewer_plugin.rs:47]: this is not blocking, could be a follow up

File: crates/bevy_render/src/view/mod.rs
Comment (robtfm) [reply crates/bevy_render/src/view/mod.rs:90]: after discussing with robswain on discord, it's fine to keep both, pls disregard this.

File: crates/bevy_pbr/src/light.rs
Comment (mockersf) [inline crates/bevy_pbr/src/light.rs:281]: ~~`num_cascades` is an size, it can't be negative~~
oh that's actually a cultural thing, in French 0 is positive but not in English

File: examples/tools/scene_viewer/scene_viewer_plugin.rs
Comment (superdump) [reply examples/tools/scene_viewer/scene_viewer_plugin.rs:47]: Good idea. We could either allow control of just the number of cascades and the first and last far bounds and have the code calculate the bounds of each cascade, or allow full control. As it is meant to be illustrative, I'm inclined to do the former as it's a bit simpler until we have some UI slider stuff.

File: crates/bevy_pbr/src/light.rs
Comment (danchia) [reply crates/bevy_pbr/src/light.rs:1707]: nope, cargo fmt puts it back that way.

File: crates/bevy_pbr/src/render/depth.wgsl
Comment (danchia) [reply crates/bevy_pbr/src/render/depth.wgsl:43]: ah thanks for showing me the convention, fixed.

File: crates/bevy_render/src/view/mod.rs
Comment (danchia) [reply crates/bevy_render/src/view/mod.rs:90]: Keeping field, but I added a comment explaining why it's optional and when you'd want to use it.

File: crates/bevy_pbr/src/render/light.rs
Comment (danchia) [reply crates/bevy_pbr/src/render/light.rs:1082]: Good catch, changed to use `num_directional_cascades_enabled` to mirror how we set it on line 1173 where we assign the `depth_texture_view` for spot light shadow map rendering.

File: examples/tools/scene_viewer/scene_viewer_plugin.rs
Comment (danchia) [reply examples/tools/scene_viewer/scene_viewer_plugin.rs:47]: sounds good, I will work on this as a follow up if that's ok with you

**Comment by danchia** [comment]
I believe feedback is addressed, with one exception (noted below). Remaining items:
* tweak `load_gltf` to look nice.
* rebase against latest master

**Comment by danchia** [comment]
I've tweaked load_gltf to shrink the distances which improves quality, and have rebased.

**Comment by rparrett** [comment]
I have a couple doc suggestions.

In a scenario like a 2.5d/3d strategy game where the camera is potentially very far away from the action, is my only option for increasing shadow map waste tweaking `nearest_bound`?

And am I right that this is effectively only pushing the end of the first cascade, potentially covering my objects with more cascades, but still wasting the entire first cascade?

Maybe this diagram will help:

```
Camera | Cascade 1          | C2       | C3  | C4 | shadow_maximum_distance
Camera | Cascade 1                     |C2 |C3 |C4| shadow_maximum_distance
                                        <<Things>>
```

1. As a user, it feels a bit unintuitive. The docs aren't incorrect or anything, but i feel like spelling out the fact that `nearest_bound` is not `minimum_shadow_distance` more explicitly might have helped me.
2. On WASM, or when there's otherwise only 1 cascade, `nearest_bound` is ignored and this doesn't seem to be mentioned anywhere.

**Comment by rparrett** [comment]
Also, is there some reason I'm missing for why we can't and/or don't really need to configure a `shadow_minimum_distance`?

**Comment by rparrett** [comment]
That makes sense, thanks. I do think the docs could be a bit clearer on this.

**Comment by superdump** [comment]
> In a scenario like a 2.5d/3d strategy game where the camera is potentially very far away from the action, is my only option for increasing shadow map waste tweaking `nearest_bound`?

`CascadeShadowConfig`'s members are public. You can construct the far bounds of the cascades however you like. It's just a `Vec<f32>`. However, the `new()` aims to be a useful convenience constructor given the usual concern of trading off performance and quality via the number of cascades and the shadow far distance. Setting the far bound of the nearest cascade is to avoid spending a bunch of cascades on very small distance ranges close to the camera.

I hear you that for a top-down/isometric type game, a different approach might indeed be useful. However, I think given there is low distribution of objects over camera depth versus distribution in screen space, the only thing that will really matter for such cases would be shadow map resolution. Then you could use a single cascade and higher shadow map texture resolution.

> And am I right that this is effectively only pushing the end of the first cascade, potentially covering my objects with more cascades, but still wasting the entire first cascade?
> 
> Maybe this diagram will help:
> 
> ```
> Camera | C1 | C2   | C3      | C4        | shadow_maximum_distance
> Camera | C1          | C2 | C3  | C4     | shadow_maximum_distance (nearest_bound tweaked)
>                      <<Things           >>
> ```

That is what adjusting the nearest bound would do, yes.

>     1. As a user, it feels a bit unintuitive. The docs aren't incorrect or anything, but i feel like spelling out the fact that `nearest_bound` is not `minimum_shadow_distance` more explicitly might have helped me.

Good point. And perhaps we could elaborate with suggestions of when to use multiple cascades and exceptions of when to use fewer or one or so.

>     2. On WASM, or when there's otherwise only 1 cascade, `nearest_bound` is ignored and this doesn't seem to be mentioned anywhere.

Also a good point, thanks for the feedback!

**Comment by superdump** [comment]
I suppose with an isometric-type view with the camera high above the action, the closest object to the camera would be at a greater distance from the camera and then as you pointed out, it would be good to not try to encompass the camera origin in the shadow mapping. Which then goes to what robtfm was saying about moving the projection's near plane, which defines the view frustum. It is probably worth someone making a test case for it and trying to put together something that works well. It's not completely clear what the recommendation would be though and I don't really want to hold up this PR. We can iterate on documentation I think.

**Review by superdump — APPROVED** [review]
@robtfm if you're ok with this then I think we can merge it.

**Review by robtfm — APPROVED** [review]
@superdump r+

**Comment by superdump** [comment]
I’ll open an issue about improving the documentation for different cases.

**Pull request closed without merging**
